warehouse/tuf/tasks.py

```diff
@@ -0,0 +1,87 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import redis
+
+from warehouse.tasks import task
+from warehouse.tuf import utils
+from warehouse.tuf.interfaces import IKeyService, IRepositoryService
+
+
+@task(bind=True, ignore_result=True, acks_late=True)
+def bump_timestamp(task, request):
+    r = redis.StrictRedis.from_url(request.registry.settings["celery.scheduler_url"])
+
+    with utils.RepoLock(r):
+        pass
+
+
+@task(bind=True, ignore_result=True, acks_late=True)
+def bump_snapshot(task, request):
+    r = redis.StrictRedis.from_url(request.registry.settings["celery.scheduler_url"])
+
+    with utils.RepoLock(r):
+        pass
+
+
+@task(bind=True, ignore_result=True, acks_late=True)
+def bump_bin_n(task, request):
+    r = redis.StrictRedis.from_url(request.registry.settings["celery.scheduler_url"])
+
+    with utils.RepoLock(r):
+        pass
+
+
+@task(bind=True, ignore_result=True, acks_late=True)
+def add_target(task, request, filepath, fileinfo):
+    r = redis.StrictRedis.from_url(request.registry.settings["celery.scheduler_url"])
+
+    with utils.RepoLock(r):
+        # TODO(ww): How slow is this? Does it make more sense to pass the loaded
+        # repository to the task?
+        repo_service = request.find_service(IRepositoryService)
+        key_service = request.find_service(IKeyService)
+        repository = repo_service.load_repository()
+
+        dirty_roles = ["snapshot", "timestamp"]
+        for role in dirty_roles:
+            role_obj = getattr(repository, role)
+            [role_obj.load_signing_key(k) for k in key_service.privkeys_for_role(role)]
+
+        # NOTE(ww): I think this should be targets("bins") instead of just targets,
```

Comment:
cc @trishankkarthik / @mnm678 / @lukpueh: I *think* this is a bug in `repository_tool.load_repository`, as each bin-n should be delegated from `bins` and not from the top-level `targets`. The repository creation code in `warehouse/cli/tuf.py` works as expected and delegates correctly from `bins`, so it only seems to happen on repo reload.

Reply:
Tracking issue: https://github.com/theupdateframework/tuf/issues/1045